AG-14318 `@SceneChangeDetection` decorator refactoring

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-14318

olegat committed:
AG-14318 Use override insteads of a switch statement

olegat committed:
AG-14318 Remove obsolete type="transform" change detection opt

olegat committed:
AG-14318 Replace ScenePathChangeDetection with SceneChangeDetection

olegat committed:
AG-14318 Remove unused SceneChangeDetectionOptions.type

olegat committed:
AG-14318 Fix TS compilation

olegat committed:
AG-14318 Do not call the changeCb twice

olegat committed:
AG-14318 Call super.onChangeDetection

This ensures that all dirty flags (_dirty, _dirtyPath and _dirtyTransform) are
all set correctly.

This fixes a bug that was reproducible on line-series. Steps to repro:
1. Open an line-series example (with 2 series & an interactive legend)
2. Toggle a line series visiblilty by clicking a legend item
   **Expected:** Marker positioning is updated.
   **Actual:** Markers positioning is not updated.

olegat committed:
AG-14318 Remove obsolete ScenePathChangeDetection lint-tests

olegat committed:
AG-14318 Use changeCb instead of override in transformable.ts

This avoids unnecessarily needing to mark the path as dirty when only a
transformation has changed.

olegat commented on packages/ag-charts-community/src/scene/changeDetectable.ts:
> [!IMPORTANT]
> This file includes most of the important changes

alantreadway replied:
As a heads-up, we'll need to validate the performance impact of these changes before merge - historically even single line changes made here have made a decent dent in performance, e.g. event something as innocuous as `changeCb?.()` only featured in one of the code-paths was a performance improvement to the others.

olegat replied:
Agreed.

I did refactor it such as that we keep the function-chaining to avoid branching-code on each set operation (https://github.com/ag-grid/ag-charts/commit/8aa8d2e0686d45c051674455a0d1af7402a37472).

I'll check what's performance impact of removing the `changeCb?.()`, but I would be surprised to see much of a difference because from the debugging that I've done it always gets call here in `type="normal"` code-paths: https://github.com/ag-grid/ag-charts/blob/77a58fe6089591bc66d9ee5bddf7d8316f2b5cd9/packages/ag-charts-community/src/scene/changeDetectable.ts#L76

olegat commented on packages/ag-charts-community/src/scene/shape/path.ts:
> [!IMPORTANT]
> This file includes important changes (removable of `ScenePathChangeDetection` and a new `onChangeDetection` override).

olegat commented on packages/ag-charts-community/src/scene/transformable.ts:
> [!IMPORTANT]
> I initially tried to implement `markDirtyTransform()` using an `onChangeDetected` override, but unfortunately this clashes with `Path._dirtyPath`).
>
> See https://github.com/ag-grid/ag-charts/commit/261c1e38e59cb78430dbf24f6d4fd6a80d29f78a

olegat committed:
AG-14318 nx format

olegat committed:
Revert "AG-14318 Use changeCb instead of override in transformable.ts"

This reverts commit 261c1e38e59cb78430dbf24f6d4fd6a80d29f78a.
This reverts commit de37fce3415018f22053debc0dafff80a252252e.

olegat committed:
AG-14318 Remove super.onChangeDetection in transformable.ts

This super-call is redundant, it's already done in the markDirtyTransform() method.

olegat committed:
Revert "AG-14318 Remove super.onChangeDetection in transformable.ts"

This reverts commit 08a01907a95d139c41397b232ba397a6d943932b.

olegat committed:
Merge remote-tracking branch 'latest' into AG-14318/SceneChangeDetection_refactor

olegat committed:
AG-14318 nx format

alantreadway reviewed (approved):
LGTM
